test/e2e/storage/persistent_volumes-local.go

```diff
@@ -1338,32 +1367,46 @@ func unmapBlockLocalVolume(config *localTestConfig, dir string, node *v1.Node) {
 	Expect(err).NotTo(HaveOccurred())
 }
 
-// Create corresponding write and read commands
-// to be executed via hostexec Pod on the node with the local PV
-func createWriteAndReadCmds(testFileDir string, testFile string, writeTestFileContent string, volumeType localVolumeType) (writeCmd string, readCmd string) {
-	writeCmd = createWriteCmd(testFileDir, testFile, writeTestFileContent, volumeType)
-	readCmd = createReadCmd(testFileDir, testFile, volumeType)
-	return writeCmd, readCmd
-}
-
-func createWriteCmd(testDir string, testFile string, writeTestFileContent string, volumeType localVolumeType) string {
-	if volumeType == BlockLocalVolumeType {
-		// testDir is the block device.
-		testFileDir := filepath.Join("/tmp", testDir)
-		testFilePath := filepath.Join(testFileDir, testFile)
-		// Create a file containing the testFileContent.
-		writeTestFileCmd := fmt.Sprintf("mkdir -p %s; echo %s > %s", testFileDir, writeTestFileContent, testFilePath)
-		// sudo is needed when using ssh exec to node.
-		// sudo is not needed and does not exist in some containers (e.g. busybox), when using pod exec.
-		sudoCmd := fmt.Sprintf("SUDO_CMD=$(which sudo); echo ${SUDO_CMD}")
-		// Write the testFileContent into the block device.
-		writeBlockCmd := fmt.Sprintf("${SUDO_CMD} dd if=%s of=%s bs=512 count=100", testFilePath, testDir)
-		// Cleanup the file containing testFileContent.
-		deleteTestFileCmd := fmt.Sprintf("rm %s", testFilePath)
-		return fmt.Sprintf("%s && %s && %s && %s", writeTestFileCmd, sudoCmd, writeBlockCmd, deleteTestFileCmd)
+func createWriteCmd(testDir string, testFile string, writeTestFileContent string, volumeType localVolumeType, whenSetUp bool) string {
+	if whenSetUp {
+		// when setup, we need to test block device command if volumeType is BlockLocalVolumeType or BlockFsWithoutFormatLocalVolumeType
+		if volumeType == BlockLocalVolumeType || volumeType == BlockFsWithoutFormatLocalVolumeType {
```

Comment:
I'm not sure if we need to treat setup specially.  Setup should create and format a fs, and then we can write a file to it just like any other fs case.

Reply:
Setup will not format block device for the new test case (do not format it manually, kubernetes will do when mounting), so when setup, we should treat it as block device, then when we create a pod, referencing it, the block device will be formatted to a specific fs type. So we should test that differently.

Reply:
Ah I see.  The read/write test cases currently expect the content to be written during setup time.

Maybe it would be better to restructure those tests to not pre-populate data during setup, and instead in the BeforeEach(), create a Pod that mounts the local PV and writes data to it.

I think that would be cleaner rather then "skipping" the test by returning success for this case, or writing a whole different set of tests just for this type.

Reply:
We will not `skipping` all the tests for the new test case
Each local volume type has several e2e test cases ,and we can already test the condition you mentioned above.
For example, in `twoPodsReadWriteTest`, we create two pods, and try to read file content immediately, but since the new test case will be treated as block device in setup, here, we will skip the reading test.  then, we will run `createWriteCmd`, pod is created now, we will write the file as we do for fs type(`mkdir, and echo filecontent to testfile`), and later `testReadFileContent` will read the file content then.
I think the test above will test all the conditions for the new test case well.
- when pod is not created(in setup), the local volume is a block device
- then, when pod is created, it will be formatted by kubernetes, we can treat it as a FS type

Reply:
Yes, the read/write will still be executed in some subset of the test cases.  I think it will be cleaner though if all test cases behave the same way across all types.

Reply:
get it, and updated